File: native/macos-fixture/Sources/SurfaceLoomMacOSFixtureModel/FixtureState.swift
import Foundation

public enum FixtureLifecycleState: String, Sendable {
    case ready
    case closing
}

public struct FixtureSnapshot: Equatable, Sendable {
    public let invokeCount: Int
    public let inputValue: String
    public let ambiguousInvocationCount: Int
    public let transientVisible: Bool
    public let lifecycleState: FixtureLifecycleState
}

public final class FixtureState: @unchecked Sendable {
    private let lock = NSLock()
    private var invokeCount = 0
    private var inputValue = ""
    private var ambiguousInvocationCount = 0
    private var transientVisible = true
    private var lifecycleState = FixtureLifecycleState.ready

    public init() {}

    @discardableResult
    public func recordInvoke() -> Int {
        lock.withLock {
            invokeCount += 1
            return invokeCount
        }
    }

    public func setValue(_ value: String) {
        lock.withLock {
            inputValue = value
        }
    }

    @discardableResult
    public func recordAmbiguousInvocation() -> Int {
        lock.withLock {
            ambiguousInvocationCount += 1
            return ambiguousInvocationCount
        }
    }

    @discardableResult
    public func dismissTransient() -> Bool {
        lock.withLock {
            guard transientVisible else { return false }
            transientVisible = false
            return true
        }
    }

    @discardableResult
    public func restoreTransient() -> Bool {
        lock.withLock {
            guard !transientVisible else { return false }
            transientVisible = true
            return true
        }
    }

    public func beginClosing() {
        lock.withLock {
            lifecycleState = .closing
        }
    }

    public func snapshot() -> FixtureSnapshot {
        lock.withLock {
            FixtureSnapshot(
                invokeCount: invokeCount,
                inputValue: inputValue,
                ambiguousInvocationCount: ambiguousInvocationCount,
                transientVisible: transientVisible,
                lifecycleState: lifecycleState
            )
        }
    }
}

File: native/windows-fixture/src/SurfaceLoom.WindowsFixture.Model/FixtureState.cs
using System.ComponentModel;
using System.Globalization;
using System.Runtime.CompilerServices;

namespace SurfaceLoom.WindowsFixture.Model;

public sealed class FixtureState : INotifyPropertyChanged
{
    private int _invokeCount;
    private string _inputValue = string.Empty;
    private int _ambiguousInvocationCount;
    private bool _isTransientVisible = true;
    private string _lifecycleState = "ready";

    public event PropertyChangedEventHandler? PropertyChanged;

    public int InvokeCount => _invokeCount;
    public string InvokeCountText => _invokeCount.ToString(CultureInfo.InvariantCulture);

    public string InputValue
    {
        get => _inputValue;
        set
        {
            ArgumentNullException.ThrowIfNull(value);
            if (_inputValue == value)
            {
                return;
            }

            _inputValue = value;
            OnPropertyChanged();
            OnPropertyChanged(nameof(ValueMirror));
        }
    }

    public string ValueMirror => _inputValue;
    public int AmbiguousInvocationCount => _ambiguousInvocationCount;
    public string AmbiguousInvocationCountText =>
        _ambiguousInvocationCount.ToString(CultureInfo.InvariantCulture);
    public bool IsTransientVisible => _isTransientVisible;
    public string LifecycleState => _lifecycleState;

    public void RecordInvoke()
    {
        checked
        {
            _invokeCount++;
        }

        OnPropertyChanged(nameof(InvokeCount));
        OnPropertyChanged(nameof(InvokeCountText));
    }

    public void RecordAmbiguousInvocation()
    {
        checked
        {
            _ambiguousInvocationCount++;
        }

        OnPropertyChanged(nameof(AmbiguousInvocationCount));
        OnPropertyChanged(nameof(AmbiguousInvocationCountText));
    }

    public bool DismissTransient()
    {
        if (!_isTransientVisible)
        {
            return false;
        }

        _isTransientVisible = false;
        OnPropertyChanged(nameof(IsTransientVisible));
        return true;
    }

    public bool RestoreTransient()
    {
        if (_isTransientVisible)
        {
            return false;
        }

        _isTransientVisible = true;
        OnPropertyChanged(nameof(IsTransientVisible));
        return true;
    }

    public void BeginClosing()
    {
        if (_lifecycleState == "closing")
        {
            return;
        }

        _lifecycleState = "closing";
        OnPropertyChanged(nameof(LifecycleState));
    }

    private void OnPropertyChanged([CallerMemberName] string? propertyName = null) =>
        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
}
